File: src/types/workflow.ts
export type NodeKind = 
  | 'start'
  | 'nmap'
  | 'http-probe'
  | 'parser'
  | 'gitleaks'
  | 'slack'
  | 'discord'
  | 'report'
  | 'delay'
  | 'condition';

export type NodeCategory = 'recon' | 'analysis' | 'security' | 'output' | 'utilities';

export interface WorkflowNode {
  id: string;
  kind: NodeKind;
  label: string;
  category: NodeCategory;
  config: Record<string, any>;
  position: { x: number; y: number };
}

export interface WorkflowEdge {
  id: string;
  source: string;
  target: string;
  sourceHandle?: string;
  targetHandle?: string;
  label?: string;
  type?: string;
}

export interface Workflow {
  id: string;
  name: string;
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
  createdAt: string;
  updatedAt: string;
  authorizedTargets: boolean;
}

export interface Finding {
  id: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  title: string;
  description: string;
  service?: string;
  port?: number;
}

export interface RunStep {
  nodeId: string;
  name: string;
  status: 'pending' | 'running' | 'succeeded' | 'failed';
  logs: string[];
  findings?: Finding[];
  startedAt?: string;
  completedAt?: string;
}

export type RunStatus = 'queued' | 'running' | 'succeeded' | 'failed';

export interface Run {
  id: string;
  workflowId: string;
  workflowName: string;
  status: RunStatus;
  startedAt: string;
  endedAt?: string;
  duration?: number;
  steps: RunStep[];
  summary?: {
    findingsCount: number;
    severities: {
      low: number;
      medium: number;
      high: number;
      critical: number;
    };
  };
}

export interface Target {
  id: string;
  value: string;
  tags: string[];
  createdAt: string;
}

export interface Integration {
  type: 'slack' | 'discord';
  webhookUrl: string;
  enabled: boolean;
}

export interface Settings {
  theme: 'light' | 'dark';
  integrations: {
    slack?: Integration;
    discord?: Integration;
  };
}

export type RunMode = 'live' | 'demo';

export interface ApiKey {
  id: string;
  name: string;
  key?: string; // Only provided on creation
  keyPrefix: string;
  createdAt: string;
  expiresAt?: string;
  lastUsed?: string;
}

File: backend/app/models/workflow.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum


class NodeKind(str, Enum):
    """Node types in workflow."""
    START = "start"
    NMAP = "nmap"
    HTTP_PROBE = "httpProbe"
    PARSER_RULES = "parserRules"
    GITLEAKS = "gitleaks"
    SLACK_ALERT = "slackAlert"
    DISCORD_ALERT = "discordAlert"
    REPORT_EXPORT = "reportExport"
    DELAY = "delay"
    CONDITION = "condition"


class NodeCategory(str, Enum):
    """Node categories."""
    UTILITY = "utility"
    RECON = "recon"
    ANALYSIS = "analysis"
    SECURITY = "security"
    OUTPUT = "output"


class Position(BaseModel):
    """Node position on canvas."""
    x: float
    y: float


class WorkflowNode(BaseModel):
    """Workflow node definition."""
    id: str
    kind: NodeKind
    label: str
    category: NodeCategory
    config: Dict[str, Any] = Field(default_factory=dict)
    position: Position


class WorkflowEdge(BaseModel):
    """Workflow edge (connection between nodes)."""
    id: str
    source: str
    target: str
    sourceHandle: Optional[str] = None
    targetHandle: Optional[str] = None
    label: Optional[str] = None
    type: Optional[str] = None


class Workflow(BaseModel):
    """Workflow model."""
    id: str
    name: str
    nodes: List[WorkflowNode]
    edges: List[WorkflowEdge]
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)
    authorizedTargets: bool = False

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class WorkflowCreate(BaseModel):
    """Create workflow request."""
    name: str
    nodes: List[WorkflowNode]
    edges: List[WorkflowEdge]
    authorizedTargets: bool = False


class WorkflowUpdate(BaseModel):
    """Update workflow request."""
    name: Optional[str] = None
    nodes: Optional[List[WorkflowNode]] = None
    edges: Optional[List[WorkflowEdge]] = None
    authorizedTargets: Optional[bool] = None
